### PR TITLE
fixes #7187 sequtils.toSeq produces the sequence from the iterator twice

### LINKED ISSUE #7187: sequtils.toSeq produces to sequence from the iterator twice if compiles(iter.len) == true
In the doc comment for sequtils.toSeq we find this example:
```nim
let
  numeric = @[1, 2, 3, 4, 5, 6, 7, 8, 9]
  odd_numbers = toSeq(filter(numeric) do (x: int) -> bool:
         if x mod 2 == 1:
           result = true)
```
If you run this, the filtered sequence is produced twice, as can be seen [here on the nim playground](https://play.nim-lang.org/?gist=7d4a355b9d440a1fbbee810a7edbf3a0) with my instrumented copies of `filter` and `toSeq`. I think its the same problem that was fixed eg. in mapIt by copying the iterable to a local variable and therefore forcing the iteration to take place.

My proposed fix looks like this:
```nim
template toSeqFixed*(iter: untyped): untyped =
  when compiles(iter.len):
    iter
  else:
    var result: seq[type(iter)] = @[]
    for x in iter:
      result.add(x)
    result
```
but, as usual, i am not sure if this fix is correct...

### PR DESCRIPTION
**Pull request by timotheecour**
* fixes #7187
* adds test for that

~~NOTE: this PR depends on https://github.com/nim-lang/Nim/pull/8584 which should be merged first~~

### CODE REVIEW

File: lib/pure/collections/sequtils.nim [hunk 1]
```diff
@@ -497,9 +513,10 @@ template toSeq*(iter: untyped): untyped =
   ##   assert odd_numbers == @[1, 3, 5, 7, 9]
 
   when compiles(iter.len):
+    evalOnce(t_D20180808T230638, iter, letAssigneable = true)
```
Comment (Varriount) [inline lib/pure/collections/sequtils.nim]: What's with the symbol name?

File: lib/pure/collections/sequtils.nim [hunk 2]
```diff
@@ -25,6 +25,22 @@ import macros
 when not defined(nimhygiene):
   {.pragma: dirty.}
 
+
+macro evalOnce(v, exp: untyped, letAssigneable: static[bool]): untyped =
+  expectKind(v, nnkIdent)
```
Comment (Varriount) [inline lib/pure/collections/sequtils.nim]: What does this do different from a single 'let' statement?

File: lib/pure/collections/sequtils.nim [hunk 3]
```diff
@@ -1106,16 +1122,12 @@ when isMainModule:
     doAssert newSeq[int](0).mapIt(it) == @[]
 
   block: # mapIt redifinition check, see https://github.com/nim-lang/Nim/issues/8580
-    let t = [1,2].mapIt(it)
-    doAssert t == @[1,2]
+    let s2 = [1,2].mapIt(it)
```
Comment (timotheecour) [inline lib/pure/collections/sequtils.nim]: `s2` is correct here as test case for https://github.com/nim-lang/Nim/issues/8580 since implementation of `mapIt` now uses s2, not t.

File: lib/pure/collections/sequtils.nim [hunk 4]
```diff
@@ -1106,16 +1122,12 @@ when isMainModule:
...
 
   block:
-    var counter = 0
-    proc getInput():auto =
-      counter.inc
-      [1, 2]
-    doAssert getInput().mapIt(it*2).mapIt(it*10) == @[20, 40]
-    # make sure argument evaluated only once, analog to
+    counter = 0
+    doAssert [1,2].identity().mapIt(it*2).mapIt(it*10) == @[20, 40]
```
Comment (timotheecour) [inline lib/pure/collections/sequtils.nim]: simplifies test case, reusing `identity` and `counter`, as done in other test cases in this file

File: lib/pure/collections/sequtils.nim [hunk 5]
```diff
@@ -1092,12 +1102,18 @@ when isMainModule:
...
+    # Corner cases (openArray litterals should not be common)
     template foo2(x: openArray[int]): seq[int] = x.mapIt(it * 10)
     counter = 0
     doAssert foo2(openArray[int]([identity(1),identity(2)])) == @[10,20]
     # TODO: this fails; not sure how to fix this case
     # doAssert counter == 2
 
+    counter = 0
+    doAssert openArray[int]([identity(1), identity(2)]).mapIt(it) == @[1,2]
+    # ditto
```
Comment (timotheecour) [inline lib/pure/collections/sequtils.nim]: added corner test case similar to the one just above

File: lib/pure/collections/sequtils.nim [hunk 6]
```diff
@@ -518,13 +518,17 @@ template toSeq*(iter: untyped): untyped =
...
+  # subtleties.
   when compiles(iter.len):
-    var i = 0
-    var result = newSeq[type(iter)](iter.len)
-    for x in iter:
-      result[i] = x
-      inc i
-    result
+    block:
+      evalOnceAs(iter2, iter, true)
```
Comment (timotheecour) [inline lib/pure/collections/sequtils.nim]: see https://github.com/nim-lang/Nim/pull/8584 for why we need evalOnceAs

File: lib/pure/collections/sequtils.nim (hunk 1 above)
Comment (timotheecour) [reply lib/pure/collections/sequtils.nim]: done

File: lib/pure/collections/sequtils.nim (hunk 2 above)
Comment (timotheecour) [reply lib/pure/collections/sequtils.nim]: done

**Comment by timotheecour** [comment]
@Varriount PTAL

**Comment by timotheecour** [comment]
@Araq could we please merge this? I have other PR's coming up depending on this
